ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala

```diff
@@ -0,0 +1,173 @@
+    val metrics = new Metrics(metricRegistry)
+    val submissionValidator = BatchedSubmissionValidator[Index](
+      BatchedSubmissionValidatorParameters.default,
+      new KeyValueCommitting(engine, metrics),
+      new ConflictDetection(metrics),
+      metrics,
+      engine,
+      NoopLedgerDataExporter
+    )
+    val inMemoryState = mutable.Map.empty[Key, Value]
+    val inMemoryLog = mutable.ArrayBuffer[LedgerRecord]()
+    val inMemoryLedgerStateOperations =
+      new InMemoryLedgerStateOperations(inMemoryLog, inMemoryState)
+    val writeRecordingLedgerStateOperations =
+      new WriteRecordingLedgerStateOperations[Index](inMemoryLedgerStateOperations)
+    val (reader, commitStrategy) =
+      BatchedSubmissionValidatorFactory.readerAndCommitStrategyFrom(
+        writeRecordingLedgerStateOperations,
+        stateKeySerializationStrategy,
+        NoopLedgerDataExporter)
+    processSubmissions(
+      input,
+      submissionValidator,
+      reader,
+      commitStrategy,
+      writeRecordingLedgerStateOperations)
+    reportDetailedMetrics(metricRegistry)
+  }
+
+  private def processSubmissions(
+      input: DataInputStream,
+      submissionValidator: BatchedSubmissionValidator[Index],
+      reader: DamlLedgerStateReader,
+      commitStrategy: CommitStrategy[Index],
+      writeRecordingLedgerStateOperations: WriteRecordingLedgerStateOperations[Index]): Unit = {
+    var counter = 0
+    while (input.available() > 0) {
+      val (submissionInfo, expectedWriteSet) = readSubmissionAndOutputs(input)
+      val validationFuture = submissionValidator.validateAndCommit(
+        submissionInfo.submissionEnvelope,
+        submissionInfo.correlationId,
+        submissionInfo.recordTimeInstant,
+        submissionInfo.participantId,
+        reader,
+        commitStrategy
+      )
+      Await.ready(validationFuture, Duration(10, TimeUnit.SECONDS))
+      counter += 1
+      val actualWriteSet = writeRecordingLedgerStateOperations.getAndClearRecordedWriteSet()
+      val sortedActualWriteSet = actualWriteSet.sortBy(_._1.asReadOnlyByteBuffer())
+      if (!compareWriteSets(expectedWriteSet, sortedActualWriteSet)) {
+        println(AnsiColor.WHITE)
+        sys.exit(1)
+      }
+    }
+    println(s"Processed $counter submissions")
+  }
+
+  private def compareWriteSets(expectedWriteSet: WriteSet, actualWriteSet: WriteSet): Boolean = {
+    if (expectedWriteSet == actualWriteSet) {
+      println(s"${AnsiColor.GREEN}OK${AnsiColor.WHITE}")
+      true
+    } else {
+      println(s"${AnsiColor.RED}failed")
+      if (expectedWriteSet.size == actualWriteSet.size) {
+        for (((expectedKey, expectedValue), (actualKey, actualValue)) <- expectedWriteSet.zip(
+            actualWriteSet)) {
+          if (expectedKey == actualKey && expectedValue != actualValue) {
+            println(
+              s"expected value: ${bytesAsHexString(expectedValue)} vs. actual value: ${bytesAsHexString(actualValue)}")
```

Comment:
Does scala string interpolator revert back to neutral color in a subsequent line after color has been switched? If not, I would suggest printing `ok` lines in neutral color and `nok` lines in red for faster analysis

Reply:
Unfortunately it doesn't revert back to neutral color, that's why I added a `println` to switch back to white before exiting. All `nok` lines are red and only `ok` is in green currently, i.e., the output looks like this:
![image](https://user-images.githubusercontent.com/57664299/85121609-5e121980-b225-11ea-993c-45281ba6d27c.png)

Reply:
That's ok, I did not read carefully enough. You only printing discrepancies to begin with.